File: VB/NamedPipeLogClient.vb
Imports DevExpress.Xpo.Logger
Imports DevExpress.Xpo.Logger.Transport
Imports System
Imports System.IO
Imports System.IO.Pipes
Imports System.Runtime.Serialization

Namespace XpoProfilerConsole
	Friend Class NamedPipeLogClient
		Implements ILogClient, IDisposable

		Private Const NamedPipeConnectTimeout As Integer = 5000
		Private ReadOnly host As Host
		Private namedPipeClient As NamedPipeClientStream

		Public Sub New(ByVal host As Host)
			Me.host = host
		End Sub

		Public Function GetCompleteLog() As LogMessage() Implements DevExpress.Xpo.Logger.Transport.ILogSource.GetCompleteLog
			Try
				EnsureCreateNamedPipe()
				namedPipeClient.WriteByte(1)
				Return GetResponseFromNamedPipe()
			Catch
				DisposeNamedPipe()
				Throw
			End Try
		End Function

		Public Function GetMessage() As LogMessage Implements DevExpress.Xpo.Logger.Transport.ILogSource.GetMessage
			Try
				EnsureCreateNamedPipe()
				namedPipeClient.WriteByte(2)
				Dim messages() As LogMessage = GetResponseFromNamedPipe()
				If messages.Length = 1 Then
					Return messages(0)
				ElseIf messages.Length = 0 Then
					Return Nothing
				End If
				Throw New InvalidOperationException("Invalid response from server.")
			Catch
				DisposeNamedPipe()
				Throw
			End Try
		End Function

		Public Function GetMessages(ByVal messageAmount As Integer) As LogMessage() Implements DevExpress.Xpo.Logger.Transport.ILogSource.GetMessages
			Try
				EnsureCreateNamedPipe()
				namedPipeClient.WriteByte(3)
				namedPipeClient.Write(BitConverter.GetBytes(messageAmount), 0, 4)
				Return GetResponseFromNamedPipe()
			Catch
				DisposeNamedPipe()
				Throw
			End Try
		End Function

		Private Function GetResponseFromNamedPipe() As LogMessage()
			Dim buf(4095) As Byte
			Dim bytesRead As Integer = namedPipeClient.Read(buf, 0, 4)
			If bytesRead <> 4 Then
				Throw New InvalidOperationException("Invalid response from server.")
			End If
			Dim dataLength As Integer = BitConverter.ToInt32(buf, 0)
			If dataLength < 0 Then
				Throw New InvalidOperationException("Invalid response from server.")
			End If
			If dataLength = 0 Then
				Return New LogMessage(){}
			End If
			Dim serializer = New DataContractSerializer(GetType(LogMessage()))
			bytesRead = 0
			Using stream = New MemoryStream()
				Do While bytesRead <> dataLength
					Dim partLength As Integer = namedPipeClient.Read(buf, 0, Math.Min(dataLength - bytesRead, buf.Length))
					If partLength = 0 OrElse Not namedPipeClient.IsConnected Then
						Throw New InvalidOperationException("Invalid response from server.")
					End If
					stream.Write(buf, 0, partLength)
					bytesRead += partLength
				Loop
				stream.Position = 0
				Return DirectCast(serializer.ReadObject(stream), LogMessage())
			End Using
		End Function

		Private Sub EnsureCreateNamedPipe()
			If namedPipeClient Is Nothing Then
				namedPipeClient = New NamedPipeClientStream(host.HostAddress, host.ServiceName, PipeDirection.InOut, PipeOptions.None, System.Security.Principal.TokenImpersonationLevel.Impersonation)
				namedPipeClient.Connect(NamedPipeConnectTimeout)
			End If
		End Sub
		Private Sub DisposeNamedPipe()
			If namedPipeClient IsNot Nothing Then
				namedPipeClient.Dispose()
				namedPipeClient = Nothing
			End If
		End Sub
		Public Function Start() As Boolean Implements ILogClient.Start
			Try
				EnsureCreateNamedPipe()
				Return True
			Catch
				Return False
			End Try
		End Function
		Public Function [Stop]() As Boolean Implements ILogClient.Stop
			DisposeNamedPipe()
			Return True
		End Function
		Public Sub Dispose() Implements System.IDisposable.Dispose
			DisposeNamedPipe()
		End Sub
	End Class
End Namespace

File: CS/NamedPipeLogClient.cs
using DevExpress.Xpo.Logger;
using DevExpress.Xpo.Logger.Transport;
using System;
using System.IO;
using System.IO.Pipes;
using System.Runtime.Serialization;

namespace XpoProfilerConsole {
    class NamedPipeLogClient : ILogClient, IDisposable {
        const int NamedPipeConnectTimeout = 5000;
        readonly Host host;
        NamedPipeClientStream namedPipeClient;

        public NamedPipeLogClient(Host host) {
            this.host = host;
        }

        public LogMessage[] GetCompleteLog() {
            try {
                EnsureCreateNamedPipe();
                namedPipeClient.WriteByte(1);
                return GetResponseFromNamedPipe();
            } catch {
                DisposeNamedPipe();
                throw;
            }
        }

        public LogMessage GetMessage() {
            try {
                EnsureCreateNamedPipe();
                namedPipeClient.WriteByte(2);
                LogMessage[] messages = GetResponseFromNamedPipe();
                if(messages.Length == 1) {
                    return messages[0];
                } else if(messages.Length == 0) {
                    return null;
                }
                throw new InvalidOperationException("Invalid response from server.");
            } catch {
                DisposeNamedPipe();
                throw;
            }
        }

        public LogMessage[] GetMessages(int messageAmount) {
            try {
                EnsureCreateNamedPipe();
                namedPipeClient.WriteByte(3);
                namedPipeClient.Write(BitConverter.GetBytes(messageAmount), 0, 4);
                return GetResponseFromNamedPipe();
            } catch {
                DisposeNamedPipe();
                throw;
            }
        }

        LogMessage[] GetResponseFromNamedPipe() {
            byte[] buf = new byte[4096];
            int bytesRead = namedPipeClient.Read(buf, 0, 4);
            if(bytesRead != 4) {
                throw new InvalidOperationException("Invalid response from server.");
            }
            int dataLength = BitConverter.ToInt32(buf, 0);
            if(dataLength < 0) {
                throw new InvalidOperationException("Invalid response from server.");
            }
            if(dataLength == 0) {
                return new LogMessage[0];
            }
            var serializer = new DataContractSerializer(typeof(LogMessage[]));
            bytesRead = 0;
            using(var stream = new MemoryStream()) {
                while(bytesRead != dataLength) {
                    int partLength = namedPipeClient.Read(buf, 0, Math.Min(dataLength - bytesRead, buf.Length));
                    if(partLength == 0 || !namedPipeClient.IsConnected) {
                        throw new InvalidOperationException("Invalid response from server.");
                    }
                    stream.Write(buf, 0, partLength);
                    bytesRead += partLength;
                }
                stream.Position = 0;
                return (LogMessage[])serializer.ReadObject(stream);
            }
        }

        void EnsureCreateNamedPipe() {
            if(namedPipeClient == null) {
                namedPipeClient = new NamedPipeClientStream(host.HostAddress, host.ServiceName, PipeDirection.InOut, PipeOptions.None, System.Security.Principal.TokenImpersonationLevel.Impersonation);
                namedPipeClient.Connect(NamedPipeConnectTimeout);
            }
        }
        void DisposeNamedPipe() {
            if(namedPipeClient != null) {
                namedPipeClient.Dispose();
                namedPipeClient = null;
            }
        }
        public bool Start() {
            try {
                EnsureCreateNamedPipe();
                return true;
            } catch {
                return false;
            }
        }
        public bool Stop() {
            DisposeNamedPipe();
            return true;
        }
        public void Dispose() {
            DisposeNamedPipe();
        }
    }
}
